docs/design-documents/drivers/serial/serial.md

```diff
@@ -0,0 +1,321 @@
+
+The relationship of `FileHandle` and other APIs is as follows:
+
+![Retarget](./diagrams/retarget.png)
+
+The POSIX layer and the static array of `FileHandle*` allocated by the retarget code for file handling consume a lot of memory. A constrained target that only uses a console does not need file handling and should be able to turn this functionality off.
+
+ A new configuration parameter is introduced to compile the file handling code out. The application can override the default configuration in mbed_app.json as follows:
+
+```C
+    "target_overrides": {
+        "*": {
+            "platform.stdio-minimal-console-only": true
+        }
+    }
+```
+
+There are two approaches to achieve a minimal retarget layer. You can:
+1. Redefine the low-level library functions. A detailed description can be found [here](http://infocenter.arm.com/help/index.jsp?topic=/com.arm.doc.dui0475m/chr1358938931411.html)
+1. Redefine the system I/O functions.
+
+A minimal console only needs to write a single character at a time. So redefining the default fputc() to directly write to the serial port if the output stream is stdout and bypassing the system I/O functions should achieve higher memory savings. If we take this approach, we will have to rework some error handlers that rely on the POSIX form of `write(STDOUT_FILENO, buf, len)` to do emergency printing.
+
+The second solution keeps the POSIX layer with the main saving coming from dropping the use of `FileHandle` and the file handle table itself. In this case, `write()` would call `MinimalConsole::putc` in a loop.
+
+List of tasks to simplify the retarget code:
+
+- Use `UnbufferedSerial` instead of `DirectSerial`
+- Remove `DirectSerial`
+- Add a configuration parameter to compile out file handling feature
+    - Alternative uses weak console_putc/getc functions
+
+## Future work-Deprecation of Stream
+
+The `Stream` class pulls in the <stdio.h> library. It should be replaced by drivers implementing `FileHandle` directly and applications using `fdopen(FileHandle *)` to get a `FILE*` to use full C/C++ stream features. 
+
+`Stream` is currently used by the following classes:
+- `Serial`
+- `USBKeyboard`
+- `USBMouseKeyboard`
+- `USBSerial`
+
+Deprecation of `Stream` should be planned in a future release.
+
+# Roll-out Plan
+
+This feature will be implemented in different phases as follows:
+
+Phase 1:
+- [Deprecation of APIs](#Deprecations)
+- Implementation of new class `UnbufferedSerial`
+- Implementation of new class `BufferedSerial`
+- Use `UnbufferedSerial` instead of `DirectSerial` in retarget code
+
+Phase 2:
+- Removal of deprecated APIs
+
+# Tools and configuration changes
+
+### Tool changes
+
+None
+
+### Configuration changes
+
+None
+
+# Other information
+
+### Deprecations
+
+The following APIs will be deprecated:
+- `Serial` class
+- `RawSerial` class
+- `UARTSerial` class
```

Comment:
DirectSerial?

Reply:
`DirectSerial` ([here](https://github.com/ARMmbed/mbed-os/blob/master/platform/source/mbed_retarget.cpp#L141).) will not be needed as `UnbufferedSerial` will be used instead.

Reply:
Doesn't need to be deprecated, as it's not public. Can just be removed.